File: macos/WaiComputer/WaiComputer/Features/Dictation/DictationStartupAudioBuffer.swift
import Foundation
import WaiComputerKit

actor DictationStartupAudioBuffer {
    enum AppendResult: Equatable {
        case buffered(chunks: Int, bytes: Int)
        case sent(bytes: Int)
    }

    enum FlushResult: Equatable {
        case flushed(chunks: Int, bytes: Int)

        var chunks: Int {
            switch self {
            case .flushed(let chunks, _): return chunks
            }
        }

        var bytes: Int {
            switch self {
            case .flushed(_, let bytes): return bytes
            }
        }
    }

    private let maxBufferedBytes: Int
    private var bufferedChunks: [Data] = []
    private var bufferedBytes = 0
    private var provider: (any ProviderSession)?
    private var isFlushing = false

    init(maxBufferedBytes: Int) {
        self.maxBufferedBytes = max(1, maxBufferedBytes)
    }

    func append(_ data: Data) async throws -> AppendResult {
        guard !data.isEmpty else {
            return .buffered(chunks: bufferedChunks.count, bytes: bufferedBytes)
        }

        if let provider, !isFlushing {
            try await provider.send(pcm16: data)
            return .sent(bytes: data.count)
        }

        let nextBufferedBytes = bufferedBytes + data.count
        guard nextBufferedBytes <= maxBufferedBytes else {
            throw DictationStartupAudioBufferError.capacityExceeded(
                bytes: nextBufferedBytes,
                limit: maxBufferedBytes
            )
        }

        bufferedChunks.append(data)
        bufferedBytes = nextBufferedBytes
        return .buffered(chunks: bufferedChunks.count, bytes: bufferedBytes)
    }

    func startStreaming(to provider: any ProviderSession) async throws -> FlushResult {
        isFlushing = true
        var sentBytes = 0
        var sentChunks = 0

        do {
            while !bufferedChunks.isEmpty {
                let chunks = bufferedChunks
                bufferedChunks = []
                bufferedBytes = 0

                for chunk in chunks {
                    try await provider.send(pcm16: chunk)
                    sentBytes += chunk.count
                    sentChunks += 1
                }
            }
        } catch {
            isFlushing = false
            self.provider = nil
            throw error
        }

        self.provider = provider
        isFlushing = false
        return .flushed(chunks: sentChunks, bytes: sentBytes)
    }

    func close() {
        provider = nil
        isFlushing = false
        bufferedChunks = []
        bufferedBytes = 0
    }
}

enum DictationStartupAudioBufferError: Error, LocalizedError, Equatable {
    case capacityExceeded(bytes: Int, limit: Int)

    var errorDescription: String? {
        switch self {
        case .capacityExceeded(let bytes, let limit):
            return "Dictation startup audio buffer exceeded \(limit) bytes (received \(bytes))."
        }
    }
}

File: desktop/WaiComputer.Core/Dictation/DictationStartupAudioBuffer.cs
namespace WaiComputer.Core.Dictation;

/// <summary>Thrown when buffered pre-roll audio exceeds the configured ceiling (no-fallback).</summary>
public sealed class DictationStartupAudioBufferException : Exception
{
    public DictationStartupAudioBufferException(int bytes, int limit)
        : base($"Dictation startup buffer overflow: {bytes} > {limit} bytes.") { }
}

/// <summary>Outcome of appending a chunk to the startup buffer.</summary>
public readonly record struct DictationAppendResult(bool WasSent, int BufferedChunks, int BufferedBytes)
{
    public static DictationAppendResult Buffered(int chunks, int bytes) => new(false, chunks, bytes);
    public static DictationAppendResult Sent() => new(true, 0, 0);
}

/// <summary>
/// Holds the pre-roll + early PCM chunks captured before the realtime provider
/// is connected, then flushes them in order once streaming starts and switches
/// to pass-through. Ports the macOS <c>DictationStartupAudioBuffer</c>:
/// capacity-limited (overflow throws — no silent drop), and serialised so an
/// append can't race the start-streaming flush.
/// </summary>
public sealed class DictationStartupAudioBuffer
{
    private readonly int _maxBufferedBytes;
    private readonly SemaphoreSlim _gate = new(1, 1);
    private readonly List<byte[]> _buffered = new();
    private int _bufferedBytes;
    private Func<ReadOnlyMemory<byte>, CancellationToken, Task>? _send;

    public DictationStartupAudioBuffer(int maxBufferedBytes) => _maxBufferedBytes = Math.Max(1, maxBufferedBytes);

    /// <summary>
    /// Append a chunk. Before streaming starts the chunk is buffered (throws if
    /// that would exceed the ceiling); once streaming is live it is sent straight
    /// through to the provider.
    /// </summary>
    public async Task<DictationAppendResult> AppendAsync(ReadOnlyMemory<byte> data, CancellationToken ct)
    {
        await _gate.WaitAsync(ct).ConfigureAwait(false);
        try
        {
            if (data.IsEmpty)
            {
                return DictationAppendResult.Buffered(_buffered.Count, _bufferedBytes);
            }

            if (_send is { } send)
            {
                await send(data, ct).ConfigureAwait(false);
                return DictationAppendResult.Sent();
            }

            var next = _bufferedBytes + data.Length;
            if (next > _maxBufferedBytes)
            {
                throw new DictationStartupAudioBufferException(next, _maxBufferedBytes);
            }

            _buffered.Add(data.ToArray());
            _bufferedBytes = next;
            return DictationAppendResult.Buffered(_buffered.Count, _bufferedBytes);
        }
        finally
        {
            _gate.Release();
        }
    }

    /// <summary>
    /// Begin streaming: flush the buffered chunks to <paramref name="send"/> in
    /// capture order, then switch to pass-through. Returns the number of chunks
    /// flushed.
    /// </summary>
    public async Task<int> StartStreamingAsync(Func<ReadOnlyMemory<byte>, CancellationToken, Task> send, CancellationToken ct)
    {
        await _gate.WaitAsync(ct).ConfigureAwait(false);
        try
        {
            var count = _buffered.Count;
            foreach (var chunk in _buffered)
            {
                await send(chunk, ct).ConfigureAwait(false);
            }
            _buffered.Clear();
            _bufferedBytes = 0;
            _send = send;
            return count;
        }
        finally
        {
            _gate.Release();
        }
    }
}
